### ISSUE TITLE
Kubelet VolumeManager incorrectly marks volume as not "in use" (i.e. safe to detach) while a mount or format is pending

### ISSUE THREAD

**Issue by saad-ali** [kubernetes/kubernetes#70319]
**What happened**:
While debugging an issue with a volume getting detached prematurely, @jingxu97 discovered an issue: the Kubelet VolumeManager incorrectly marks volume as not "in use" (i.e. safe to detach) while a mount or format is pending.

For context, attach/detach controller will not proceed with a detach (after pod referencing a volume is deleted) while the volume is still marked as "in use" by kubelet. Kubelet marks a volume as "in use" before it starts mounting or formating the volume.

The problem is:

* If a pod referencing a volume is deleted while a mount (or format) is pending, then kubelet will go ahead and remove that pod from its "desired state of the world" cache (normally this waits for pod's containers to terminate, but in this case the containers never even started because mount is still pending, i.e. format, is still pending).
* Kubelet volume_manager's `GetVolumesInUse()` method (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L293) reports the volume as no longer in use because it is not in desired or actual state (it does not check pending operations <-- this is the bug).
* Attach/detach controller sees the pod delete, checks and sees the volume is not marked as "in use" and proceeds with the detach (incorrectly), but a mount (or format) operation may be pending.

**What you expected to happen**:

Kubelet volume_manager's `GetVolumesInUse()` method (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L293) should count volumes that have pending operations (mount or otherwise) as "in use" (even if they are not in actual or desired state).

**How to reproduce it (as minimally and precisely as possible)**:

While mount is pending (because format is taking a long time, for example), delete the pod referencing the attachable volume.

**Anything else we need to know?**:
/sig storage
/milestone 1.13

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

**Comment by saad-ali** [comment]
Fix for this should be cherry picked back to all recent supported versions.

**Comment by mlmhl** [comment]
It seems that we need to mark all attached volumes in `ActualStateOfWorld` as `in-use`, not only globally mounted volumes.

We can add a `GetInUseVolumes` method to ASW and return all volumes exist in `ActualStateOfWorld.attachedVolumes`.

**Comment by saad-ali** [comment]
We don't need all volumes to be reported, only attachable volumes. Each "in use" volume ends up in the node object, so if we report every volume we would just unnecessarily increase the size of the node object. And it wouldn't fix the root cause here, because a volume could be not yet in `ActualStateOfWorld` and also removed from `DesiredStateOfWorld` but have a pending operation.

Best way to fix this is to change Kubelet volume_manager's `GetVolumesInUse()` method (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L293) to look at volumes in pending operations in addition to actual/desired state when figuring out what is "in use".

**Comment by msau42** [comment]
@kubernetes/sig-storage-bugs

**Comment by AishSundar** [comment]
@saad-ali and @msau42 is there an owner for this issue?

**Comment by msau42** [comment]
@gnufied would you be able to help investigate this issue?

**Comment by pohly** [comment]
Is there ever a situation where there should be more than one operation in flight per volume?

This seems risky. How about serializing all changes per volume (both long-running operations and those direct functions) inside `reconcile()`? This is stricter than the "consider volumes as in use when there are pending operations" and would give me more confidence that the code really does the right thing under all circumstances.

**Comment by nikopen** [comment]
Pinged @gnufied via PM, leaving this here for visibility. 
Tracking closely due to priority

**Comment by AishSundar** [comment]
@gnufied @saad-ali @msau42 
Is there any progress on this issue? Its a critical-urgent one. Thanks

**Comment by msau42** [comment]
If we addressed https://github.com/kubernetes/kubernetes/issues/69831, would this first event not happen?

> If a pod referencing a volume is deleted while a mount (or format) is pending, then kubelet will go ahead and remove that pod from its "desired state of the world" cache (normally this waits for pod's containers to terminate, but in this case the containers never even started because mount is still pending, i.e. format, is still pending).

**Comment by nikopen** [comment]
@AishSundar It's a storage bug affecting all recent k8s versions which can be patched, it doesn't have to do anything in particular with features of the next release. If it doesn't make it it will likely be included in patch v1.13.1

@jingxu97 Indeed it's not affected by code freeze since it's a bug. I think last merge window before v1.13 is around end of November IIRC.

**Comment by AishSundar** [comment]
@jingxu97 @nikopen Unfortunately Code freeze means a blanket freeze for any kind of PR into k/k unless its highly critical and affects the stability of 1.13.

If we think this issue is not a major blocker foe 1.13 and we can live with it until 1.13.1, RT will not be approving this into master post tomorrow's freeze.

@saad-ali @msau42 to clarify the severity and priority of this having to go in 1.13 vs 1.13.1. Thanks

**Comment by msau42** [comment]
Fixing this in a patch release is fine.  This is an issue that has existed in kubernetes for many releases now.

**Comment by gnufied** [comment]
@saad-ali @msau42 as I was saying - the bug is not in the path you described, it exists somewhere else. Before we start mount operation for an attachable volume, via `VerifyControllerAttachedVolume` we add the volume to actual state of the world of volume manager. So even if pod was deleted before mount could finish - the volume still is *in actual_state_of_world* and hence should be reported as `in-use` by the node, because that function considers both desired and actual state of the world - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L314

**Comment by mlmhl** [comment]
> the volume still is in actual_state_of_world and hence should be reported as in-use by the node, because that function considers both desired and actual state of the world

Currently `VolumeManager` only considers globally mounted volumes as `in-use`: https://github.com/kubernetes/kubernetes/blob/a856c7ab1de1bed266cbeab5dec703105d1bf2d0/pkg/kubelet/volumemanager/volume_manager.go#L298, but `VerifyControllerAttachedVolume` sets `globallyMounted` to `false` when adding the volume to `ASW`: https://github.com/kubernetes/kubernetes/blob/a856c7ab1de1bed266cbeab5dec703105d1bf2d0/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L404, so after `VerifyControllerAttachedVolume` finished, volume is still considered as `not in-use`.

In my previous comments, I suggested to consider all attached volumes in `ASW` as `in-use`, not only globally mounted volumes, but I'm not sure if it covers all situations.

**Comment by gnufied** [comment]
@mlmhl yeah you are correct. This is basically same as issue @jingxu97 logged awhile back - https://github.com/kubernetes/kubernetes/issues/40887 . 

I was able to reproduce this issue by introducing  a `time.Sleep` in Format step.

**Comment by gnufied** [comment]
PR to fix the issue -https://github.com/kubernetes/kubernetes/pull/71145
